ext/yarp/extension.c

```diff
@@ -302,16 +302,65 @@ lex_file(VALUE self, VALUE filepath) {
 /* Parsing Ruby code                                                          */
 /******************************************************************************/
 
+// This is passed as a callback to the parser. It gets called every time a new
+// token is found from within a call to parse that accepted a block.
+static void
+parse_token(void *data, yp_parser_t *parser, yp_token_t *token) {
+    lex_data_t *lex_data = (lex_data_t *) parser->lex_callback->data;
+    rb_yield_values(2, yp_token_new(parser, token, lex_data->encoding, lex_data->source), INT2FIX(parser->lex_state));
+}
+
 // Parse the given input and return a ParseResult instance.
 static VALUE
 parse_input(yp_string_t *input, const char *filepath) {
     yp_parser_t parser;
     yp_parser_init(&parser, yp_string_source(input), yp_string_length(input), filepath);
 
+    VALUE offsets;
+    VALUE source;
+
+    // If a block was given to the parse method, then we're going to register a
+    // lex callback that will yield the tokens to the block. This means you can
+    // get the lexer and the parser output in one method call instead of having
+    // to parse twice.
+    if (rb_block_given_p()) {
```

Comment:
BTW does the parser gem needs tokens and nodes, not just nodes?

Reply:
The parser gem doesn't need tokens, but rubocop does. I don't really know why.

I've updated it to return `[node, tokens]` in a parse result instead of a block, and it works with the FFI now.

Reply:
Nice!